Photocopier Improvements

timothymtorres opened the pull request:
## About The Pull Request
Photocopiers no longer initialize all their stored papers. Since the only paper that can be stored has to be empty this is now just stored as a number value. This prevents hundreds of blank papers from being spawned during round start loading, since every station map has a dozen of these machines scattered about.

Other stuff:
- Emag function where it breaks the payment system so people are no longer charged for printing
- Stock parts now affect max paper storage, print time, and toner usage (swapped servo part for matter bin)
- Can now load paper via paper bins
- Can now select which type of paper to print with (regular, carbon, construction, natural)
- Fix stored paper counts being wrong
- Gave paper subtypes proper names so we can use them in our UI
- Paper and toner are now dropped when machine deconstructed
- Refactored item code to use `item_interaction()` instead of `attackby()`

## Why It's Good For The Game
New functionality, bugfixes, refactored code, and optimizations are good.

## Changelog
:cl:
add: Photocopiers can now be emagged to disable the payment system that charges you for printing copies
add: Photocopiers are now affected by stock part upgrades. Matter bins affect paper capacity, lasers affect toner usage, and scanning module affects time to print. The servo stock part was replaced with matter bin.
add: Photocopiers can now be loaded with regular, carbon, construction, and natural paper. You can select which kind to use in the UI when printing copies if any are stored.
qol: Photocopiers can now be loaded with paper bins
fix: Photocopiers now display the correct amount of stored paper
fix: Photocopiers now drop toner and paper when machine is deconstructed
code: Paper subtypes (carbon, construction, natural) were given proper names to use with photocopiers UI menu.
refactor: Photocopier code was refactored to not spawn paper when it is initialized. Instead it stores blank papers as a numeric value. Also switched the `attackby()` code handling to use the modern `item_interaction()` proc.
/:cl:

timothymtorres committed:
Refactor a ton of bad photocopier code

timothymtorres committed:
Add rainbow icon for construction paper and proper name

timothymtorres committed:
Setup photocopier for robustness

timothymtorres committed:
Add name to paper subtypes

timothymtorres committed:
Add selectable paper types for photocopiers

timothymtorres committed:
Switch servo stock part with matter bin

timothymtorres committed:
Add typescript UI support for paper menu

timothymtorres committed:
Add stock part improvments and emag ability

timothymtorres committed:
Merge branch 'master' into fix-photocopier-paper

timothymtorres committed:
Remove deprecated code

timothymtorres committed:
Make sure user can insert item

timothymtorres committed:
Update code comment

timothymtorres committed:
Remove deprecated code comment

timothymtorres committed:
Fix typescript error

timothymtorres committed:
Fix some more redundant code

00-Steven commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -439,7 +550,7 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
 		return null
 	var/obj/item/photo/copied_pic = new(src, photo.Copy(photo_color == PHOTO_GREYSCALE ? TRUE : FALSE))
 	delete_paper(PHOTO_PAPER_USE)
-	toner_cartridge.charges -= PHOTO_TONER_USE
+	toner_cartridge.charges -= PHOTO_TONER_USE / toner_efficiency
```
I see a lot of repeated
```dm
toner_cartridge.charges -= BASE_AMOUNT_DEFINE / toner_efficiency
```
maybe better to move it under a `use_toner(base_amount)` proc, so the calculation for toner efficiency is only in one place

timothymtorres replied:
Applied via https://github.com/tgstation/tgstation/pull/91258/commits/7c92685df75204ad97c4a068a17c12d5539dba6d

00-Steven commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -546,47 +657,111 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
-				return
-		insert_copy_object(object, user)
+/obj/machinery/photocopier/item_interaction(mob/living/user, obj/item/tool, list/modifiers)
+	//if(!user.temporarilyRemoveItemFromInventory(tool))
+	//	return ITEM_INTERACT_FAILURE
```
oops commented out code in the pr

timothymtorres replied:
Removed via https://github.com/tgstation/tgstation/pull/91258/commits/23e4339b372e5cc56ec92083847289f17603319b

00-Steven reviewed (commented):
only did a quick scan, haven't read through the full thing in detail yet

I'm a bit iffy on making it purely based on paper types and just deleting empty paper and adding to the stack, cause paper being empty doesn't actually mean it's not distinct from newly created paper of its type- like this means spraypainted colours and forensics are lost, which *practically* isn't very important, it just feels kinda off

timothymtorres committed:
Remove commented out code

Loafin34 commented:
Emagging the photocopier should shoot out evil paper planes at people until it runs out of ammo: the price decrease is a bit too weak imo.

timothymtorres committed:
Refactor toner use into proc

timothymtorres commented:
> I'm a bit iffy on making it purely based on paper types and just deleting empty paper and adding to the stack, cause paper being empty doesn't actually mean it's not distinct from newly created paper of its type- like this means spraypainted colours and forensics are lost, which _practically_ isn't very important, it just feels kinda off

There are a lot of objects that have this behavior. Putting wires or glass into machines? Deleted. Pipes wrenched off the floor? Deleted. The paper that is inserted into the photocopier is meant to be clean/fresh paper. It can't be written on, crumbled, cut, or folded. For all intents and purposes, it's basically getting deleted to be reused as matter.

Gaxeer commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -97,35 +117,73 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
+/obj/machinery/photocopier/RefreshParts()
+	. = ..()
+	max_paper_capacity = 0
+	for(var/datum/stock_part/matter_bin/matter_bin in component_parts)
+		max_paper_capacity += MATTER_BIN_PAPER_CAPACITY * matter_bin.tier
+
+	toner_efficiency = 1
+	for(var/datum/stock_part/micro_laser/micro_laser in component_parts)
+		toner_efficiency += micro_laser.tier
+
+	time_to_print = 5 SECONDS
+	for(var/datum/stock_part/scanning_module/scanning_module in component_parts)
+		time_to_print -= (scanning_module.tier SECONDS)
```
you iterate same list 3 times, instead of doing typechecks during 1 iteration like this:
```
max_paper_capacity = 0
toner_efficiency = 1
time_to_print = 5 SECONDS
for(var/datum/stock_part/part as anything in component_parts)
	if(istype(part, /datum/stock_part/matter_bin)
		max_paper_capacity += MATTER_BIN_PAPER_CAPACITY * matter_bin.tier
	else if(istype(part, /datum/stock_part/micro_laser)
		toner_efficiency += micro_laser.tier
	else if(istype(part, /datum/stock_part/scanning_module)
		time_to_print -= (scanning_module.tier SECONDS)
```

timothymtorres replied:
Good catch. I actually reused this code from other machines RefreshParts() proc so it can probably be applied in many places.

timothymtorres replied:
> you iterate same list 3 times, instead of doing typechecks during 1 iteration like this:

![chrome_Ef0WEGFPmI](https://github.com/user-attachments/assets/632289df-dc52-4ecf-bcc9-8f8164a6132b)

Just a heads up, you can use the "Add a suggestion" button to suggest code edits that will get commited to the PR and attribute you as a co-author. Your code in the current form is broken though because you need to do `part.tier` instead of `matter_bin.tier`, `micro_laser.tier`, `scanning_module.tier`

timothymtorres replied:
Looks like this is a bad idea:
- https://github.com/tgstation/tgstation/pull/91301

00-Steven commented:
> > I'm a bit iffy on making it purely based on paper types and just deleting empty paper and adding to the stack, cause paper being empty doesn't actually mean it's not distinct from newly created paper of its type- like this means spraypainted colours and forensics are lost, which _practically_ isn't very important, it just feels kinda off
> 
> There are a lot of objects that have this behavior. Putting wires or glass into machines? Deleted. Pipes wrenched off the floor? Deleted. The paper that is inserted into the photocopier is meant to be clean/fresh paper. It can't be written on, crumbled, cut, or folded. For all intents and purposes, it's basically getting deleted to be reused as matter.

Ah yeah, it's just that currently one could insert pre-coloured paper into the photocopier and print onto that, and this limits it to instead those specific papers
It's not super important though so, shrug. If ethanol cleaning maintained stamps and you could print onto stamped paper I'd have stronger opinions, but such is't real (~~_yet_~~)

MrMelbert commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -297,30 +382,40 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
+		copies_amount = FLOOR(get_paper_count(created_paper) / paper_use, 1)
 		error_message = span_warning("An error message flashes across \the [src]'s screen: \"Not enough paper to perform [copies_amount >= 1 ? "full " : ""]operation.\"")
-	if(copies_amount > 0 && (attempt_charge(src, user, (copies_amount - 1) * PHOTOCOPIER_FEE) & COMPONENT_OBJ_CANCEL_CHARGE))
+	if(!(obj_flags % EMAGGED) && (copies_amount > 0) && (attempt_charge(src, user, (copies_amount - 1) * PHOTOCOPIER_FEE) & COMPONENT_OBJ_CANCEL_CHARGE))
```
```suggestion
	if(!(obj_flags & EMAGGED) && (copies_amount > 0) && (attempt_charge(src, user, (copies_amount - 1) * PHOTOCOPIER_FEE) & COMPONENT_OBJ_CANCEL_CHARGE))
```

timothymtorres committed:
Fix emagged flag

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

timothymtorres committed:
Add whitespace to argument

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

timothymtorres committed:
Add more whitespace to arguments

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

timothymtorres committed:
Add on_deconstruction method to photocopier

timothymtorres commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -97,35 +117,73 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
 	if(object_copy)
 		object_copy.forceMove(drop_location())
 
+	dump_contents()
```
Applied via https://github.com/tgstation/tgstation/pull/91258/commits/8f1ba5f8e9bd1eea7ca38f6c18ab8bfde76d81ee

timothymtorres committed:
Refactor object copy drop locatoin for destroy and dump

timothymtorres committed:
Fix grammer for method name

timothymtorres commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -546,47 +656,108 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
 
-/obj/machinery/photocopier/proc/insert_copy_object(obj/item/object, mob/user)
+/// Check if there is enough room to insert paper
+/obj/machinery/photocopier/proc/is_room_for_paper(mob/user, amount=1)
```
Applied via https://github.com/tgstation/tgstation/pull/91258/commits/8c5ca5b7186ad110abbc25cb219e8419a14138a8

timothymtorres committed:
Refactor toner cartidge removal

timothymtorres committed:
Fix toner cartridge runtime

timothymtorres committed:
Fix paper teleporting after copying

timothymtorres committed:
Merge branch 'master' into fix-photocopier-paper

Ghommie committed:
Merge branch 'master' into fix-photocopier-paper
